File: programmers/이분탐색/징검다리 건너기.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/64062
#include <string>
#include <vector>

using namespace std;
int N, maxL, answer = 0;

int solution(vector<int> stones, int k) {
    N = stones.size();
    for(int i = 0; i < N; i++){
        maxL = max(maxL, stones[i]);
    }
    int s = 1, e = maxL;
    while(s <= e){
        int mid = (s + e)/2;
        
        int cnt = 0, maxCnt = 0;
        for(int i = 0; i < N; i++){
            if(stones[i] <= mid){ //  0 이하가 나오는 개수
                cnt++;
            }else{
                maxCnt = max(maxCnt, cnt);
                cnt = 0;
            }
        }
        maxCnt = max(maxCnt, cnt);
        
        if(k <= maxCnt){    // 사람을 너무 많이 뺐다
            e = mid - 1;
        }else{
            s = mid + 1;
            answer = max(answer, s); //주의
        }
    }
    
    return answer;
}

File: python/이분탐색/징검다리 건너기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/64062?language=python3
def solution(stones, k):
    answer = 0
    # 건널 수 있는 최소 인원 1명, 최대 인원은 가장 큰 돌의 숫자
    s, e = 1, max(stones)
    while s <= e:
        mid = (s+e) // 2    # 이번에 건너볼 인원 수
        
        cnt = 0
        check = True
        for stone in stones:
            if stone < mid:
                cnt += 1
            else:
                cnt = 0
            
            if cnt >= k :   # 못 밟는 돌이 연속으로 k개 이상 나오면 mid명은 못 건넘!
                check = False
                break
                
        if check:
            answer = mid
            s = mid + 1
        else:
            e = mid - 1
            
    return answer